packages/autorest.python/autorest/codegen/serializers/model_serializer.py

```diff
@@ -266,21 +275,26 @@ def declare_property(prop: Property) -> str:
         field = "rest_discriminator" if prop.is_discriminator else "rest_field"
         type_ignore = (
             prop.is_discriminator
-            and prop.is_discriminator
-            and cast(ConstantType, prop.type).value
+            and isinstance(prop.type, (ConstantType, EnumValueType))
+            and prop.type.value
         )
+
         return (
-            f"{prop.client_name}: {prop.type_annotation()} ="
+            f"{prop.client_name}: {_type_annotation(prop)} ="
             f' {field}({", ".join(args)}){"  # type: ignore" if type_ignore else ""}'
         )
 
     def initialize_properties(self, model: ModelType) -> List[str]:
         init_args = []
         for prop in self.get_properties_to_declare(model):
             if prop.constant or prop.is_discriminator:
+                declaration = (
+                    prop.type.get_declaration()
+                    if isinstance(prop.type, (ConstantType, EnumValueType))
+                    else None
+                )
```

Comment:
a discriminator property should be either a constant or an enum value. can we simplify this logic? it seems else is useless.

Reply:
Discriminator type of base model can be `constant/enum` and discriminator type of sub model types can be `constant/enum/enumvalue`